platforms/js/Native.hx

```diff
@@ -1749,36 +1749,146 @@ class Native {
-				for (i in 0...fields.length)
-					fields[i] = HaxeRuntime.fastMakeStructValue2("Pair", fields[i], object2JsonStructs( untyped __js__ ("o[fields[i]]") ));
-				return HaxeRuntime.fastMakeStructValue("JsonObject", fields);
+			var t = typeof o;
+			switch (t) {
+				case 'string':
+					var obj = { _id : Native.sidJsonString};
+					obj[Native.sidJsonStringVal] = o;
+					return obj;
+				case 'number':
+					var obj = { _id : Native.sidJsonDouble};
+					obj[Native.sidJsonDoubleVal] = o;
+					return obj;
+				case 'boolean':
+					var obj = { _id : Native.sidJsonBool};
+					obj[Native.sidJsonBoolVal] = o;
+					return obj;
+				default:
+					if(o == null) {
+						return Native.jsonNull;
+					} else {
+						var mappedFields = [];
+						for(var f in o) {
+							var a2 = Native.object2JsonStructs(o[f]);
+							var obj = { _id : Native.sidPair};
+							obj[Native.sidPairFirst] = f;
+							obj[Native.sidPairSecond] = a2;
+							mappedFields.push(obj);
+						}
+						var obj = { _id : Native.sidJsonObject};
+						obj[Native.sidJsonObjectFields] = mappedFields;
+						return obj;
+					}
 			}
-		}
+		}");
+
+		return "";
+	}
+
+	// Firefox and maybe other browsers faster with Object.getOwnPropertyNames that with for(var f in o)
+	private static function object2JsonStructs_FF(o : Dynamic) : Dynamic {
+		untyped __js__("
+		if (Array.isArray(o)) {
+			var a1 = Native.map(o,Native.object2JsonStructs_FF);
+			var obj = { _id : Native.sidJsonArray};
+			obj[Native.sidJsonArrayVal] = a1;
+			return obj;
+		} else {
+			var t = typeof o;
+			switch (t) {
+				case 'string':
+					var obj = { _id : Native.sidJsonString};
+					obj[Native.sidJsonStringVal] = o;
+					return obj;
+				case 'number':
+					var obj = { _id : Native.sidJsonDouble};
+					obj[Native.sidJsonDoubleVal] = o;
+					return obj;
+				case 'boolean':
+					var obj = { _id : Native.sidJsonBool};
+					obj[Native.sidJsonBoolVal] = o;
+					return obj;
+				default:
+					if(o == null) {
+						return Native.jsonNull;
+					} else {
+						var mappedFields = Object.getOwnPropertyNames(o);
+						for(var i=0; i< mappedFields.length; i++) {
+							var f = mappedFields[i];
+							var a2 = Native.object2JsonStructs_FF(o[f]);
+							var obj = { _id : Native.sidPair};
+							obj[Native.sidPairFirst] = f;
+							obj[Native.sidPairSecond] = a2;
+							mappedFields[i] = obj;
+						}
+						var obj = { _id : Native.sidJsonObject};
+						obj[Native.sidJsonObjectFields] = mappedFields;
+						return obj;
+					}
+			}
+		}");
+
+		return "";
 	}
 
+	private static var parseJsonFirstCall = true;
 	public static function parseJson(json : String) : Dynamic {
-	 	try {
-	 		var o = haxe.Json.parse(json);
-			return object2JsonStructs(o);
-		} catch (e : Dynamic) {
-			return makeStructValue("JsonDouble", [0.0], null); 
-		}
+		try {
+			if (parseJsonFirstCall) {
+				untyped Native.sidJsonArray = HaxeRuntime._structids_.h["JsonArray"];
```

Comment:
it should not rely on Haxe hashmap inner implementation and should use .get() here

Reply:
addressed